File: rust/linkedspec-runtime/src/semantic_observation.rs
//! Invocation-local typed semantic observations for native Rust execution.

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::cell::RefCell;
use std::fmt;
use std::fmt::Write;
use std::rc::Rc;

/// Versioned identity carried by every native semantic observation event.
pub const RUNTIME_SEMANTIC_OBSERVATION_CONTRACT: &str =
    "linkedspec-semantic-execution-observation-v1";

/// Closed event vocabulary captured by the v1 Rust runtime observer.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RuntimeSemanticObservationEventKind {
    RegexSlotSelected,
    RuleResult,
}

/// One clone-safe native event delivered during normal parser execution.
///
/// Fields remain public so callers can persist the typed observation. The
/// semantic index validates the complete schema and topology before accepting
/// any caller-retained event sequence.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RuntimeSemanticObservationEvent {
    /// Versioned schema identity for this event.
    pub contract_id: String,
    /// Closed semantic event kind.
    pub event_kind: RuntimeSemanticObservationEventKind,
    /// Rule executing at the observation seam.
    pub rule_label: String,
    /// Authored rule that owns a selected regex slot, for slot events only.
    pub target_rule: Option<String>,
    /// Zero-based regex slot index within `target_rule`, for slot events only.
    pub regex_index: Option<usize>,
    /// Unicode-scalar cursor position immediately after selection or completion.
    pub position: usize,
    /// SHA-256 identity of the exact UTF-8 input bytes, for result events only.
    pub input_identity: Option<String>,
    /// Stable completion status, for result events only.
    pub status: Option<String>,
}

impl RuntimeSemanticObservationEvent {
    pub(crate) fn regex_slot_selected(
        rule_label: &str,
        target_rule: &str,
        regex_index: usize,
        position: usize,
    ) -> Self {
        Self {
            contract_id: RUNTIME_SEMANTIC_OBSERVATION_CONTRACT.to_string(),
            event_kind: RuntimeSemanticObservationEventKind::RegexSlotSelected,
            rule_label: rule_label.to_string(),
            target_rule: Some(target_rule.to_string()),
            regex_index: Some(regex_index),
            position,
            input_identity: None,
            status: None,
        }
    }

    pub(crate) fn rule_result(rule_label: &str, position: usize, input: &str) -> Self {
        Self {
            contract_id: RUNTIME_SEMANTIC_OBSERVATION_CONTRACT.to_string(),
            event_kind: RuntimeSemanticObservationEventKind::RuleResult,
            rule_label: rule_label.to_string(),
            target_rule: None,
            regex_index: None,
            position,
            input_identity: Some(input_identity(input)),
            status: Some("succeeded".to_string()),
        }
    }
}

type SinkCallback = dyn FnMut(RuntimeSemanticObservationEvent);

/// Caller-owned handle installed into one execution through `ExecutionOptions`.
///
/// Clones share the same callback, which lets the caller retain one handle while
/// the invocation-local runtime context owns another. The callback is
/// deliberately infallible: a caller panic unwinds unchanged and is never
/// translated into a parser, trace, or diagnostic failure.
#[derive(Clone)]
pub struct RuntimeSemanticObservationSink {
    callback: Rc<RefCell<Box<SinkCallback>>>,
}

impl RuntimeSemanticObservationSink {
    /// Wrap a typed synchronous event callback.
    pub fn new<F>(callback: F) -> Self
    where
        F: FnMut(RuntimeSemanticObservationEvent) + 'static,
    {
        Self {
            callback: Rc::new(RefCell::new(Box::new(callback))),
        }
    }

    pub(crate) fn emit(&self, event: RuntimeSemanticObservationEvent) {
        (self.callback.borrow_mut())(event);
    }
}

impl fmt::Debug for RuntimeSemanticObservationSink {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("RuntimeSemanticObservationSink")
            .finish_non_exhaustive()
    }
}

impl PartialEq for RuntimeSemanticObservationSink {
    fn eq(&self, other: &Self) -> bool {
        Rc::ptr_eq(&self.callback, &other.callback)
    }
}

impl Eq for RuntimeSemanticObservationSink {}

fn input_identity(input: &str) -> String {
    let digest = Sha256::digest(input.as_bytes());
    let mut identity = String::with_capacity("input:sha256:".len() + digest.len() * 2);
    identity.push_str("input:sha256:");
    for byte in digest {
        write!(&mut identity, "{byte:02x}").expect("writing to String cannot fail");
    }
    identity
}

File: lua/src/linkedspec/semantic_observation.lua
-- Invocation-local typed runtime semantic observation for the Lua backend.
-- Event construction and sink-failure carriers stay package-internal; callers
-- receive protected immutable handles and detached JSON projections.

local json = require("linkedspec.json")
local sha256 = require("linkedspec.sha256")

local M = {}

M.CONTRACT_ID = "linkedspec-semantic-execution-observation-v1"
M.REGEX_SLOT_SELECTED = "regex_slot_selected"
M.RULE_RESULT = "rule_result"

local EVENT_STATE = setmetatable({}, { __mode = "k" })
local SINK_FAILURE_STATE = setmetatable({}, { __mode = "k" })

local function empty_pairs()
  return function() return nil end, nil, nil
end

local function immutable_newindex()
  error("Runtime semantic observation events are immutable", 0)
end

local EVENT_MT = {
  __index = function(value, key)
    local state = EVENT_STATE[value]
    return state == nil and nil or state[key]
  end,
  __newindex = immutable_newindex,
  __pairs = empty_pairs,
  __metatable = "protected",
  __tostring = function() return "RuntimeSemanticObservationEvent" end,
}

local SINK_FAILURE_MT = {
  __metatable = "protected",
  __tostring = function() return "RuntimeSemanticObservationSinkFailure" end,
}

local function event(fields, use_default_contract)
  local result = setmetatable({}, EVENT_MT)
  EVENT_STATE[result] = {
    contract_id = use_default_contract and M.CONTRACT_ID or fields.contract_id,
    event_kind = fields.event_kind,
    rule_label = fields.rule_label,
    target_rule = fields.target_rule,
    regex_index = fields.regex_index,
    position = fields.position,
    input_identity = fields.input_identity,
    status = fields.status,
  }
  return result
end

-- Package-private malformed-event constructor for derivation boundary tests.
-- The root module deliberately does not export this seam.
function M._event_for_testing(fields)
  if type(fields) ~= "table" then error("event fields must be a table", 0) end
  return event(fields, false)
end

function M.regex_slot_selected(rule_label, target_rule, regex_index, position)
  return event({
    event_kind = M.REGEX_SLOT_SELECTED,
    rule_label = rule_label,
    target_rule = target_rule,
    regex_index = regex_index,
    position = position,
  }, true)
end

function M.rule_result(rule_label, position, input)
  return event({
    event_kind = M.RULE_RESULT,
    rule_label = rule_label,
    position = position,
    input_identity = "input:sha256:" .. sha256.hex(input),
    status = "succeeded",
  }, true)
end

function M.is_event(value)
  return type(value) == "table" and EVENT_STATE[value] ~= nil
end

function M.to_json(value)
  local state = EVENT_STATE[value]
  if state == nil then error("expected RuntimeSemanticObservationEvent", 0) end
  return json.harray({
    contract_id = state.contract_id,
    event_kind = state.event_kind,
    rule_label = state.rule_label,
    target_rule = state.target_rule == nil and json.null or state.target_rule,
    regex_index = state.regex_index == nil and json.null or state.regex_index,
    position = state.position,
    input_identity = state.input_identity == nil and json.null or state.input_identity,
    status = state.status == nil and json.null or state.status,
  })
end

function M.deliver(sink, value)
  local delivered, failure = pcall(sink, value)
  if delivered then return end
  local carrier = setmetatable({}, SINK_FAILURE_MT)
  SINK_FAILURE_STATE[carrier] = { value = failure }
  error(carrier, 0)
end

function M.is_sink_failure(value)
  return type(value) == "table" and SINK_FAILURE_STATE[value] ~= nil
end

function M.sink_failure_value(value)
  local state = SINK_FAILURE_STATE[value]
  return state == nil and nil or state.value
end

return M
